src/camera.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright © 2018-2019 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gio/gunixfdlist.h>
+#include <gio/gdesktopappinfo.h>
+#include <stdio.h>
+
+#include "device.h"
+#include "request.h"
+#include "permissions.h"
+#include "pipewire.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+
+typedef struct _Camera Camera;
+typedef struct _CameraClass CameraClass;
+
+struct _Camera
+{
+  XdpCameraSkeleton parent_instance;
+};
+
+struct _CameraClass
+{
+  XdpCameraSkeletonClass parent_class;
+};
+
+static Camera *camera;
+
+GType camera_get_type (void);
+static void camera_iface_init (XdpCameraIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (Camera, camera, XDP_TYPE_CAMERA_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_TYPE_CAMERA,
+                                                camera_iface_init))
+
+static void
+handle_access_camera_in_thread_func (GTask *task,
+                                     gpointer source_object,
+                                     gpointer task_data,
+                                     GCancellable *cancellable)
+{
+  Request *request = (Request *)task_data;
+  const char *app_id;
+  const char *device;
+  gboolean allowed;
+  g_autoptr(GError) error = NULL;
+
+  REQUEST_AUTOLOCK (request);
+
+  app_id = (const char *)g_object_get_data (G_OBJECT (request), "app-id");
+  device = (const char *)g_object_get_data (G_OBJECT (request), "device");
```

Comment:
Seems a little silly to funnel the string "camera" through object data when the function is explicitly called handle_access_camera_...

I would just hardcode "camera" here, in the device_query_permission call, like you do in handle_open_pipewire_remote